lib/src/back_end/solution.dart

```diff
@@ -130,17 +138,26 @@ class Solution implements Comparable<Solution> {
     _selectionEnd = end;
   }
 
-  /// Mark this solution as having a newline where none is permitted and thus
-  /// not a valid solution.
+  /// Mark this solution as having a newline where none is permitted by [piece]
+  /// and is thus not a valid solution.
   ///
   /// This should only be called by [CodeWriter].
-  void invalidate() {
+  void invalidate(Piece piece) {
     _isValid = false;
+
+    // If the piece whose newline constraint was violated is already bound to
+    // one state, then every solution derived from this one will also fail in
+    // the same way, so discard the whole solution tree hanging off this one.
+    if (isBound(piece)) _canExpandSolution = false;
```

Comment:
Nice, we love to see it.

Reply:
It took me a _long_ time to figure out why I couldn't just use `isValid` for this. So glad I was able to figure it out.